src/System.Management.Automation/utils/PasteArguments.cs

```diff
@@ -0,0 +1,158 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+// Fetched from https://github.com/dotnet/runtime/blob/a3befe1917526e5c735cd8a528cbee280fd66df2/src/libraries/System.Private.CoreLib/src/System/PasteArguments.cs
+
+using System.Collections.Generic;
+using System.Text;
+
+namespace System.Management.Automation.Internal
+{
+    /// <summary>
+    /// Utility class for quoting arguments according to MSVCRT rules, pre- or post- 2008.
+    /// The same rules are used by CoreFX for cmdline parsing too.
+    /// </summary>
+    public static class PasteArguments
+    {
+        /// <summary>
+        /// Append one argument to the cmdline being built, without separator.
+        /// </summary>
+        /// <remarks>
+        /// This version can be used incrementally for mixing quoted and unquoted runs.
+        /// </remarks>
+        /// <param name="stringBuilder">The cmdline being built.</param>
+        /// <param name="argument">The argument to append.</param>
+        /// <param name="forceQuote">When true, force the argument to be quoted. Normally only empty ones and those with spaces or quotes are quoted. May help avoid globbing.</param>
+        public static void AppendArgument(StringBuilder stringBuilder, string argument, bool forceQuote)
+        {
+            // Parsing rules for non-argv[0] arguments:
+            //   - Backslash is a normal character except followed by a quote.
+            //   - 2N backslashes followed by a quote ==> N literal backslashes followed by unescaped quote
+            //   - 2N+1 backslashes followed by a quote ==> N literal backslashes followed by a literal quote
+            //   - Parsing stops at first whitespace outside of quoted region.
+            //   - (post 2008 rule): A closing quote followed by another quote ==> literal quote, and parsing remains in quoting mode.
+            if (!forceQuote && argument.Length != 0 && IsLiteralSafe(argument))
+            {
+                // Simple case - no quoting or changes needed.
+                stringBuilder.Append(argument);
+            }
+            else
+            {
+                // Sanitize for incremental use.
+                if (stringBuilder[stringBuilder.Length - 1] == Quote)
+                {
+                    // Already quoting. Do not add a quote; just reuse by trimming the existing one.
+                    // "abc""def" is different from "abcdef". We want the latter.
+                    stringBuilder.Length -= 1;
+                }
+                else
+                {
+                    // If the stuff is unquoted, we need to double the trailing backslashes since we
+                    // are putting a quote here.
+                    int numBackSlash = 0;
+                    for (int i = stringBuilder.Length - 1; i > 0 && stringBuilder[i] == Backslash; i--)
```

Comment:
StringBuilder really just holds a stack of chunks with the last thing on top. So like appends, this is quite efficient given we are doing stuff on the tail.

(I was worrying about ending up on some sort of "accidentally O(n log n)" page.)